File: server/matchmaking/src/main/java/com/boxtrotstudio/ghost/matchmaking/network/packets/MatchRedirectPacket.java
package com.boxtrotstudio.ghost.matchmaking.network.packets;

import com.boxtrotstudio.ghost.matchmaking.core.hosts.gameserver.GameServer;
import com.boxtrotstudio.ghost.matchmaking.network.PlayerClient;
import com.boxtrotstudio.ghost.matchmaking.network.TcpServer;
import com.boxtrotstudio.ghost.network.packet.Packet;

import java.io.IOException;

public class MatchRedirectPacket extends Packet<TcpServer, PlayerClient> {
    public MatchRedirectPacket(PlayerClient client) {
        super(client);
    }

    @Override
    public void onWritePacket(PlayerClient client, Object... args) throws IOException {
        String ip;
        int port;

        if (args.length == 1) {
            GameServer server = (GameServer) args[0];

            ip = server.getIp();
            port = server.getPort();
        } else {
            ip = (String)args[0];
            port = (short)args[1];
        }

        write((byte)0x26)
                .write((byte)ip.length())
                .write(ip)
                .write(port)
                .endTCP();
    }
}

File: client/core/src/com/boxtrotstudio/ghost/client/network/packets/MatchRedirectPacket.kt
package com.boxtrotstudio.ghost.client.network.packets

import com.badlogic.gdx.Gdx
import com.boxtrotstudio.ghost.client.Ghost
import com.boxtrotstudio.ghost.client.handlers.GameHandler
import com.boxtrotstudio.ghost.client.network.Packet
import com.boxtrotstudio.ghost.client.network.PlayerClient
import java.io.IOException

class MatchRedirectPacket : Packet<PlayerClient>() {
    override fun handle() {
        val ip = consume(consume(1).asByte().toInt()).asString()
        val port = consume(2).asShort()

        Ghost.client = PlayerClient.connect(ip + ":" + port)
        val packet = SessionPacket()

        packet.writePacket(Ghost.client, Ghost.Session);
        if (!Ghost.client.ok()) {
            throw IOException("Bad session!");
        }
        Ghost.client.isValidated = true

        Gdx.app.postRunnable {
            Ghost.getInstance().clearScreen()
            val game = GameHandler(ip, Ghost.Session)
            game.start()
            Ghost.getInstance().handler = game
        }
    }
}
